src/languages/javascript.js

```diff
@@ -24,12 +24,26 @@ export default function(hljs) {
     literal: ECMAScript.LITERALS.join(" "),
     built_in: ECMAScript.BUILT_INS.join(" ")
   };
+  const nonDecimalLiterals = (prefixLetters, validChars) => 
+    `(-?)\\b0[${prefixLetters}][${validChars}]([${validChars}_]*[${validChars}])?n?`;
+  const noLeadingZeroDecimalDigits = /[1-9]([0-9_]*\d)?/;
+  const decimalDigits = /\d([0-9_]*\d)?/;
+  const exponentPart = regex.concat(/[eE][+-]?/, decimalDigits);
   var NUMBER = {
     className: 'number',
     variants: [
-      { begin: '\\b(0[bB][01]+)n?' },
-      { begin: '\\b(0[oO][0-7]+)n?' },
-      { begin: hljs.C_NUMBER_RE + 'n?' }
+      { begin: nonDecimalLiterals('bB', '01') }, // Binary literals
+      { begin: nonDecimalLiterals('oO', '0-7') }, // Octal literals
+      { begin: nonDecimalLiterals('xX', '0-9a-fA-F') }, // Hexadecimal literals
+      { begin: regex.concat(/(-?)\b/, noLeadingZeroDecimalDigits, 'n') }, // Non-zero BigInt literals
+      { begin: regex.concat(/(-?)0?\./, decimalDigits, regex.optional(exponentPart)) }, // Decimal literals between 0 and 1
+      { begin: regex.concat(
+        /(-?)\b/,
+        noLeadingZeroDecimalDigits,
+        regex.optional(regex.concat(/\./, regex.optional(decimalDigits))), // fractional part
+        regex.optional(exponentPart)
+        ) }, // Decimal literals >= 1
+      { begin: /(-?)\b0[\.n]?/ }, // Zero literals (`0`, `0.`, `0n`, `-0`, ...)
```

Comment:
Is there a strong reason you have included `-`?  Typically we do not do this because it will break math expressions like `3-2`... there is no simple way to detect that it's truly a unary expression or not...

Sorry for the long delay on this... finally polished up my TS/JS work and hope to have it merged soon so I was taking another look at this.

Reply:
> Is there a strong reason you have included `-`?

TBH I have no recollection what I was trying to achieve here. Once the other PR gets merge, I'll try to come back at this one to see if that's OK to just remove that dash.

Reply:
> Once the other PR gets merge,

Merged. :)